templates/commands/describe/describe.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package describe implements the template describe related subcommands.
+package describe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"path/filepath"
+	"text/tabwriter"
+
+	"github.com/abcxyz/abc/templates/common"
+	"github.com/abcxyz/abc/templates/common/templatesource"
+	"github.com/abcxyz/abc/templates/model/decode"
+	spec "github.com/abcxyz/abc/templates/model/spec/v1beta1"
+	"github.com/abcxyz/pkg/cli"
+)
+
+type Command struct {
+	cli.BaseCommand
+	flags DescribeFlags
+
+	testFS common.FS
+}
+
+const (
+	// The spec file is always located in the template root dir and named spec.yaml.
+	specName = "spec.yaml"
+)
+
+// Desc implements cli.Command.
+func (c *Command) Desc() string {
+	return "describe a template base on it's spec.yaml"
+}
+
+func (c *Command) Help() string {
+	return `
+Usage: {{ COMMAND }} [options] <source>
+
+The {{ COMMAND }} command describe the given template.
+
+The "<source>" is the location of the template to be instantiated. Many forms
+are accepted:
+
+  - "helloworld@v1" means "github.com/abcxyz/helloworld repo at revision v1;
+    this is for a template owned by abcxyz.
+  - "myorg/helloworld@v1" means github.com/myorg/helloworld repo at revision
+    v1; this is for a template not owned by abcxyz but still on GitHub.
+  - "mygithost.com/mygitrepo/helloworld@v1" is for a template in a remote git
+    repo but not owned by abcxyz and not on GitHub.
+  - "mylocaltarball.tgz" is for a template not in git but present on the local
+    filesystem.
+  - "http://example.com/myremotetarball.tgz" os for a non-Git template in a
+    remote tarball.`
+}
+
+func (c *Command) Flags() *cli.FlagSet {
+	set := c.NewFlagSet()
+	c.flags.Register(set)
+	return set
+}
+
+type runParams struct {
+	fs     common.FS
+	stdout io.Writer
+
+	describeTempDirBase string
+}
+
+func (c *Command) Run(ctx context.Context, args []string) error {
+	if err := c.Flags().Parse(args); err != nil {
+		return fmt.Errorf("failed to parse flags: %w", err)
+	}
+	fSys := c.testFS
+	if fSys == nil {
+		fSys = &common.RealFS{}
+	}
+	return c.realRun(ctx, &runParams{
+		fs:     fSys,
+		stdout: c.Stdout(),
+	})
+}
+
+// readRun provides fakable interface to test Run.
+func (c *Command) realRun(ctx context.Context, rp *runParams) (rErr error) {
+	var tempDirs []string
+	defer func() {
+		var rErr error
```

Comment:
I don't think you want this line

Reply:
The function signature already has this as a named return parameter, so this is creating a local shadow of the same variable name.

Reply:
Ah, right.